Find out the injection point in testing environment where we can hook this agent

Typical test framework takes advantage of reflection to execute stateless tests. With this observation, demo shown in https://github.com/r30shah/openj9-jit-debug-agent#openj9-jit-debug-agent injects the agent to trigger the debug agent tool in `java/lang/reflect/Method.invoke`.
In typical test suite, there might be number of tests where throwing an exception would be expected behaviour, and injecting the agent at the `Method.invoke` in such cases would lead to false positives. Find out how typical test framework executes such cases, and find out a reliable injection point so that we can weed-out possible false positives.

@qasimy123  Thanks a lot for the logs. Interesting. In my environment with following Unit tests, 
[NewTest.zip](https://github.com/r30shah/openj9-jit-debug-agent/files/7565593/NewTest.zip) and printing the the stack trace where the agent is triggered,

```
Expected Exception NULL
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/com.ibm.jit.JITHelpers.invoke(JITHelpers.java:1194)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:62)
	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:49)
	at DebugAgentRunner.run(DebugAgentRunner.java:24)
Caused by: java.lang.NullPointerException
	at Test_String_Modified.test_Constructor13(Test_String_Modified.java:201)
	... 33 more
Caught java.lang.NullPointerException inside JITHelpers
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/com.ibm.jit.JITHelpers.invoke(JITHelpers.java:1194)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:62)
	at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:49)
	at DebugAgentRunner.run(DebugAgentRunner.java:24)
Caused by: java.lang.NullPointerException
	at Test_String_Modified.test_Constructor13(Test_String_Modified.java:201)
	... 33 more
```

If you see the Test_String_Modified.test_Constructor13 in the attached unit test, you can see that it is expected to throw NPE. Can you try this in your environment? 

Please keep in mind, this is with the bug introduced by `TR_DebugAgentBug` as I am looking to prepare a sample multi-threaded unit test that can demonstrate the use of this agent.

--- a/jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java
+++ b/jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java
@@ -50,6 +50,7 @@ public final class JITHelpers {
 
 	private static final JITHelpers helpers;
 	private static final Unsafe unsafe;
+	private static java.util.concurrent.ConcurrentHashMap<String, Class<? extends Throwable>> exceptionMap;
 
 	private JITHelpers() {
 	}
@@ -67,6 +68,50 @@ public final class JITHelpers {
 	 */
 	private static JITHelpers jitHelpers() {
 		return helpers;
+	}
+
+	/**
+	 * Set the expected exception for the current thread.
+	 * Initializes the exceptionMap if necessary.
+	 * 
+	 * @param expectedException The class of the expected exception.
+	 */
+
+	public static void setExpectedException(Class<? extends Throwable> expectedException) {
+		Thread currentThread = Thread.currentThread();
+		String vmThreadName = currentThread.getName();
+		initExceptionMap();
+		
+		if (expectedException == null) {
+			exceptionMap.remove(vmThreadName);
+		}
+		else {
+			exceptionMap.put(vmThreadName, expectedException);
+		}
+	}
+
+	/**
+	 * Get the expected exception for the current thread.
+	 * 
+	 * @return The class of the expected exception.
+	 */
+
+	public static Class<? extends Throwable> getExpectedException() {
+		Thread currentThread = Thread.currentThread();
+		if(exceptionMap == null) {
+			return null;
+		}
+		return exceptionMap.get(currentThread.getName());
+	}
+
+	/**
+	 * Synchronized method to instantiate an exceptionMap if it does not exist
+	 */
+
+	public synchronized static void initExceptionMap() {
+		if (exceptionMap == null) {
+			exceptionMap = new java.util.concurrent.ConcurrentHashMap<String, Class<? extends Throwable>>();
+		}
 	}
 
 	public native int transformedEncodeUTF16Big(long src, long dest, int num);
@@ -1180,18 +1225,44 @@ public final class JITHelpers {
 
 	private native static final void debugAgentRun(MethodAccessor ma, Object obj, Object[] args);
 
+	/**
+	 * Invokes the method on the object with the given MethodAccessor and arguments.
+	 * If the method throws an exception, it is caught and if the exception is unexpected,
+	 * the debug agent triggered.
+	 * 
+	 * @param ma the method to run the debug agent on.
+	 * @param obj the underlying object.
+	 * @param args the arguments for the method
+	 * @return the return value of the method
+	 */
 	public static Object invoke(MethodAccessor ma, Object obj, Object[] args) throws InvocationTargetException {
 		try {
 			return ma.invoke(obj, args);
 		} catch (InvocationTargetException e) {
 			if (e.getCause() != null && e.getCause().getClass().getName().equals("java.lang.NullPointerException")) {
-				// TODO: Need synchronization to prevent many threads entering here
-				System.err.println("Caught java.lang.NullPointerException inside JITHelpers");
-
-				debugAgentRun(ma, obj, args);
+				Class < ? extends Throwable > expected = com.ibm.jit.JITHelpers.getExpectedException();
+				boolean runDebugAgent = true;
+				if (expected != null) {
+					Throwable targetException = e.getTargetException();
+					if (!expected.isAssignableFrom(targetException.getClass())) {
+						String message = "From JITHelpers.invoke - Unexpected exception, expected<" +
+							expected.getName() + "> but was<" +
+							targetException.getClass().getName() + ">";
+						System.out.println(message);
+					} else {
+						runDebugAgent = false;
+					}
+				}
+				if (runDebugAgent) {
+					synchronized (JITHelpers.class) {
+						System.err.println("Caught java.lang.NullPointerException inside JITHelpers, thread "+Thread.currentThread().getName());
+						e.getCause().printStackTrace();
+						debugAgentRun(ma, obj, args);
 				
-				System.err.println("Aborting JVM");
-				System.exit(1);
+						System.err.println("Aborting JVM");
+						System.exit(1);
+					}
+				}
 			}
 
 			throw e;

